framework/src/actions/CreateExecutionerAction.C

```diff
@@ -41,26 +44,46 @@ CreateExecutionerAction::CreateExecutionerAction(InputParameters params)
 void
 CreateExecutionerAction::act()
 {
-  std::shared_ptr<EigenProblem> eigen_problem = std::dynamic_pointer_cast<EigenProblem>(_problem);
-  if (eigen_problem)
-    _moose_object_pars.set<EigenProblem *>("_eigen_problem") = eigen_problem.get();
+  if (_current_task == "setup_executioner")
+  {
+    std::shared_ptr<EigenProblem> eigen_problem = std::dynamic_pointer_cast<EigenProblem>(_problem);
+    if (eigen_problem)
+      _moose_object_pars.set<EigenProblem *>("_eigen_problem") = eigen_problem.get();
 
-  std::shared_ptr<Executioner> executioner =
+    std::shared_ptr<Executioner> executioner =
       _factory.create<Executioner>(_type, "Executioner", _moose_object_pars);
 
-  std::shared_ptr<Eigenvalue> eigen_executioner =
+    std::shared_ptr<Eigenvalue> eigen_executioner =
       std::dynamic_pointer_cast<Eigenvalue>(executioner);
 
-  if ((eigen_problem == nullptr) != (eigen_executioner == nullptr))
-    mooseError("Executioner is not consistent with each other; EigenExecutioner needs an "
-               "EigenProblem, and Steady and Transient need a FEProblem");
+    if ((eigen_problem == nullptr) != (eigen_executioner == nullptr))
+      mooseError("Executioner is not consistent with each other; EigenExecutioner needs an "
+                 "EigenProblem, and Steady and Transient need a FEProblem");
+
+    // If enabled, automatically create a Preconditioner if the [Preconditioning] block is not found
+    if (_auto_preconditioning && !_awh.hasActions("add_preconditioning") &&
+        _moose_object_pars.isParamValid("solve_type"))
+      setupAutoPreconditioning();
 
-  // If enabled, automatically create a Preconditioner if the [Preconditioning] block is not found
-  if (_auto_preconditioning && !_awh.hasActions("add_preconditioning") &&
-      _moose_object_pars.isParamValid("solve_type"))
-    setupAutoPreconditioning();
+    _app.setExecutioner(std::move(executioner));
+  }
+  else if (_current_task == "setup_executioner_complete")
+  {
+    if (!isParamValid("solve_object_ordering"))
+      return;
 
-  _app.setExecutioner(std::move(executioner));
+    auto exe = _app.getExecutioner();
+    auto & ordering = getParam<std::vector<std::string>>("solve_object_ordering");
+    if (ordering.size() == 0)
+      mooseError("'solve_object_ordering' cannot be empty if it is given in the input");
+    exe->setHeadSolveObject(ordering[0]);
+    for (unsigned int i = 1; i < ordering.size(); ++i)
+    {
+      auto lead_object = exe->getSolveObject(ordering[i - 1]);
+      auto current_object = exe->getSolveObject(ordering[i]);
+      lead_object->setInnerSolve(*current_object);
+    }
+  }
```

Comment:
so either this or setting inner_solve from the input block of the solve objects. 
This is less flexible. I m fine with it if it works for the most complex griffin case desired

Reply:
There actually will be another way, where derived executioners can check the constructed solve objects and chain them with solve objects added by the executioners. The executioners expects certain types of solve objects either constructed or not. I think we can live with this for now. Overall if solve objects are developed properly, we can change this infrastructure easily in the future.

Reply:
i dont like this other way because it means steady/transient/eigenvalue will have to duplicate code to add their chain of solve objects

Reply:
@friedmud how serious is the need for multiple inner solves? This design disallows it pretty hard